### ISSUE TITLE
MacOS current master unable to fetch certificates and/or properly set KeyUsage flags

### ISSUE THREAD

**Issue by mouse07410** [OpenSC/OpenSC#1030]
### Expected behaviour

At the very least, `pkcs11-tool -O -l` should display the correct KeyUsage flags. Like this:
```
$ pkcs11-tool -O -l
Using slot 1 with a present token (0x4)
Logging in to "PIV Card Holder pin (PIV_II)".
Please enter User PIN: 
Private Key Object; RSA 
  label:      PIV AUTH key
  ID:         01
  Usage:      decrypt, sign, unwrap
Public Key Object; RSA 2048 bits
  label:      PIV AUTH pubkey
  ID:         01
  Usage:      encrypt, verify, wrap
Certificate Object, type = X.509 cert
  label:      Certificate for PIV Authentication
  ID:         01
Private Key Object; RSA 
  label:      SIGN key
  ID:         02
  Usage:      decrypt, sign, non-repudiation
  Access:     always authenticate
Public Key Object; RSA 2048 bits
  label:      SIGN pubkey
  ID:         02
  Usage:      encrypt, verify
Certificate Object, type = X.509 cert
  label:      Certificate for Digital Signature
  ID:         02
Private Key Object; RSA 
  label:      KEY MAN key
  ID:         03
  Usage:      decrypt, unwrap
Public Key Object; RSA 2048 bits
  label:      KEY MAN pubkey
  ID:         03
  Usage:      encrypt, wrap
Certificate Object, type = X.509 cert
  label:      Certificate for Key Management
  ID:         03
Data object 1121998000
  label:          'Cardholder Fingerprints'
  application:    'Cardholder Fingerprints'
  app_id:         2.16.840.1.101.3.7.2.96.16
  flags:           private
Data object 1121998096
  label:          'Printed Information'
  application:    'Printed Information'
  app_id:         2.16.840.1.101.3.7.2.48.1
  flags:           private
Data object 1121998192
  label:          'Cardholder Facial Image'
  application:    'Cardholder Facial Image'
  app_id:         2.16.840.1.101.3.7.2.96.48
  flags:           private
Private Key Object; RSA 
  label:      CARD AUTH key
  ID:         04
  Usage:      sign
Public Key Object; RSA 2048 bits
  label:      CARD AUTH pubkey
  ID:         04
  Usage:      verify
Certificate Object, type = X.509 cert
  label:      Certificate for Card Authentication
  ID:         04
Data object 1122006080
  label:          'Card Capability Container'
  application:    'Card Capability Container'
  app_id:         2.16.840.1.101.3.7.1.219.0
  flags:          <empty>
Data object 1122003776
  label:          'Card Holder Unique Identifier'
  application:    'Card Holder Unique Identifier'
  app_id:         2.16.840.1.101.3.7.2.48.0
  flags:          <empty>
Data object 1122003872
  label:          'Unsigned Card Holder Unique Identifier'
  application:    'Unsigned Card Holder Unique Identifier'
  app_id:         2.16.840.1.101.3.7.2.48.2
  flags:          <empty>
Data object 1122003968
  label:          'X.509 Certificate for PIV Authentication'
  application:    'X.509 Certificate for PIV Authentication'
  app_id:         2.16.840.1.101.3.7.2.1.1
  flags:          <empty>
Data object 1121998288
  label:          'X.509 Certificate for Digital Signature'
  application:    'X.509 Certificate for Digital Signature'
  app_id:         2.16.840.1.101.3.7.2.1.0
  flags:          <empty>
Data object 1121998384
  label:          'X.509 Certificate for Key Management'
  application:    'X.509 Certificate for Key Management'
  app_id:         2.16.840.1.101.3.7.2.1.2
  flags:          <empty>
Data object 1121998480
  label:          'X.509 Certificate for Card Authentication'
  application:    'X.509 Certificate for Card Authentication'
  app_id:         2.16.840.1.101.3.7.2.5.0
  flags:          <empty>
Data object 1121998576
  label:          'Security Object'
  application:    'Security Object'
  app_id:         2.16.840.1.101.3.7.2.144.0
  flags:          <empty>
Data object 1121998672
  label:          'Discovery Object'
  application:    'Discovery Object'
  app_id:         2.16.840.1.101.3.7.2.96.80
  flags:          <empty>
Data object 1121998768
  label:          'Key History Object'
  application:    'Key History Object'
  app_id:         2.16.840.1.101.3.7.2.96.96
  flags:          <empty>
$
```

### Actual behaviour

KeyUsage flags are mis-placed. So, for example, `KEY MAN key` gets `Usage: Sign`, which is hardly proper for the Key Management key.

```
$ env DYLD_LIBRARY_PATH=${DYLD_LIBRARY_PATH}:/tmp/lib PKCS11SPY_OUTPUT=/tmp/spy.log.txt /tmp/bin/pkcs11-tool --module /tmp/lib/pkcs11-spy.so -O -l
Using slot 0 with a present token (0x0)
Logging in to "PIV Card Holder pin (PIV_II)".
Please enter User PIN: 
Private Key Object; RSA 
  label:      PIV AUTH key
  ID:         01
  Usage:      decrypt, unwrap
Public Key Object; RSA 2048 bits
  label:      PIV AUTH pubkey
  ID:         01
  Usage:      encrypt, wrap
Certificate Object; type = X.509 cert
  label:      Certificate for PIV Authentication
  ID:         01
Private Key Object; RSA 
  label:      SIGN key
  ID:         02
  Usage:      decrypt, sign, unwrap
  Access:     always authenticate
Public Key Object; RSA 2048 bits
  label:      SIGN pubkey
  ID:         02
  Usage:      encrypt, verify, wrap
Certificate Object; type = X.509 cert
  label:      Certificate for Digital Signature
  ID:         02
Private Key Object; RSA 
  label:      KEY MAN key
  ID:         03
  Usage:      sign
Public Key Object; RSA 2048 bits
  label:      KEY MAN pubkey
  ID:         03
  Usage:      verify
Certificate Object; type = X.509 cert
  label:      Certificate for Key Management
  ID:         03
Data object 293634720
  label:          'Cardholder Fingerprints'
  application:    'Cardholder Fingerprints'
  app_id:         2.16.840.1.101.3.7.2.96.16
  flags:           private
Data object 293634816
  label:          'Printed Information'
  application:    'Printed Information'
  app_id:         2.16.840.1.101.3.7.2.48.1
  flags:           private
Data object 293634912
  label:          'Cardholder Facial Image'
  application:    'Cardholder Facial Image'
  app_id:         2.16.840.1.101.3.7.2.96.48
  flags:           private
Private Key Object; RSA 
  label:      CARD AUTH key
  ID:         04
  Usage:      decrypt, unwrap
Public Key Object; RSA 2048 bits
  label:      CARD AUTH pubkey
  ID:         04
  Usage:      encrypt, wrap
Certificate Object; type = X.509 cert
  label:      Certificate for Card Authentication
  ID:         04
Data object 293636768
  label:          'Card Capability Container'
  application:    'Card Capability Container'
  app_id:         2.16.840.1.101.3.7.1.219.0
  flags:          <empty>
Data object 293630256
  label:          'Card Holder Unique Identifier'
  application:    'Card Holder Unique Identifier'
  app_id:         2.16.840.1.101.3.7.2.48.0
  flags:          <empty>
Data object 293630352
  label:          'Unsigned Card Holder Unique Identifier'
  application:    'Unsigned Card Holder Unique Identifier'
  app_id:         2.16.840.1.101.3.7.2.48.2
  flags:          <empty>
Data object 293630448
  label:          'X.509 Certificate for PIV Authentication'
  application:    'X.509 Certificate for PIV Authentication'
  app_id:         2.16.840.1.101.3.7.2.1.1
  flags:          <empty>
Data object 293635008
  label:          'X.509 Certificate for Digital Signature'
  application:    'X.509 Certificate for Digital Signature'
  app_id:         2.16.840.1.101.3.7.2.1.0
  flags:          <empty>
Data object 293635104
  label:          'X.509 Certificate for Key Management'
  application:    'X.509 Certificate for Key Management'
  app_id:         2.16.840.1.101.3.7.2.1.2
  flags:          <empty>
Data object 293635200
  label:          'X.509 Certificate for Card Authentication'
  application:    'X.509 Certificate for Card Authentication'
  app_id:         2.16.840.1.101.3.7.2.5.0
  flags:          <empty>
Data object 293635296
  label:          'Security Object'
  application:    'Security Object'
  app_id:         2.16.840.1.101.3.7.2.144.0
  flags:          <empty>
Data object 293635392
  label:          'Discovery Object'
  application:    'Discovery Object'
  app_id:         2.16.840.1.101.3.7.2.96.80
  flags:          <empty>
Data object 293635488
  label:          'Key History Object'
  application:    'Key History Object'
  app_id:         2.16.840.1.101.3.7.2.96.96
  flags:          <empty>
$
```

### Steps to reproduce

1. Clone from Github on Mac.
2. Run `OpenSC Uninstaller.app` to ensure no remnants from the previous working installation is still around.
3. Build and install, e.g., via `./bootstrap && MacOSX/build` (assuming #1029 doesn't stop you).
4. Insert a token, and try `pkcs11-tool -O -l`. Observe the output.

### Logs

The main OpenSC debug log:
[opensc-master-bad.txt](https://github.com/OpenSC/OpenSC/files/935362/opensc-master-bad.txt)

SPY capture: 
[spy.log.txt](https://github.com/OpenSC/OpenSC/files/935367/spy.log.txt)

**Comment by mouse07410** [comment]
Bisecting... At this time, 81d16fc is OK. Continuing...

**Comment by mouse07410** [comment]
And here's the patch to revert 777e2a3 because git did not revert it cleanly: 
[opensc-remove-cac.zip](https://github.com/OpenSC/OpenSC/files/938522/opensc-remove-cac.zip)

**Comment by frankmorgner** [comment]
No problem, but please don't produce so much noise while searching for a problem 😉 

In your log the PIV card driver was used. From 777e2a3 only the changes in `pkcs15-cert.c` should be relevant:
- https://github.com/OpenSC/OpenSC/commit/777e2a3751e3f6d53f056c98e9e20e42af674fb1#diff-ab0d446b4ca95d307f37fbcebe1eac84L319
- https://github.com/OpenSC/OpenSC/commit/777e2a3751e3f6d53f056c98e9e20e42af674fb1#diff-ab0d446b4ca95d307f37fbcebe1eac84R328

Please try reverting the two individual lines if that helps. @Jakuje , do you have an idea what is going wrong?

**Comment by dengert** [comment]
The difference is in  sc_pkcs15_get_bitstring_extension where the CAC code changed SC_ASN1_BIT_STRING to SC_ASN1_BIT_STRING_NI

 -		{ "bitString", SC_ASN1_BIT_STRING, SC_ASN1_TAG_BIT_STRING, 0, value, &val_len },
 +		{ "bitString", SC_ASN1_BIT_STRING_NI, SC_ASN1_TAG_BIT_STRING, 0, value, &val_len },

Why was this change made for the CAC? It broke running code. 

See RFC 5280 and PKCS#11 V2.40 Note that there may be as many as 9 bits in the keyUsage,  
and more could be added in the future. 
 
You really don't want the NI, because in ASN1 DER bits are added on the end.

**Comment by mouse07410** [comment]
@dengert thank you!!

> The difference is in `sc_pkcs15_get_bitstring_extension()` where the CAC code changed `SC_ASN1_BIT_STRING` to `SC_ASN1_BIT_STRING_NI`. 
> Why was this change made for the CAC? It broke running code.

I confirm. Reverting that one definition from `SC_ASN1_BIT_STRING_NI` back to `SC_ASN1_BIT_STRING` seems to have restored the working code. Current OpenSC master now assigns to usage flags reasonable values (excerpts from RSA and ECC tokens):
```
. . . . .
Private Key Object; RSA 
  label:      SIGN key
  ID:         02
  Usage:      decrypt, sign, non-repudiation, unwrap
  Access:     always authenticate
Public Key Object; RSA 2048 bits
  label:      SIGN pubkey
  ID:         02
  Usage:      encrypt, verify, wrap
Certificate Object; type = X.509 cert
  label:      Certificate for Digital Signature
  ID:         02
Private Key Object; RSA 
  label:      KEY MAN key
  ID:         03
  Usage:      decrypt, unwrap
Public Key Object; RSA 2048 bits
  label:      KEY MAN pubkey
  ID:         03
  Usage:      encrypt, wrap
. . . . .
Private Key Object; EC
  label:      SIGN key
  ID:         02
  Usage:      sign, non-repudiation
  Access:     always authenticate
Public Key Object; EC  EC_POINT 256 bits
  EC_POINT:   0441xxxxxx
  EC_PARAMS:  0608xxxxxx
  label:      SIGN pubkey
  ID:         02
  Usage:      verify
Certificate Object; type = X.509 cert
  label:      Certificate for Digital Signature
  ID:         02
Private Key Object; EC
  label:      KEY MAN key
  ID:         03
  Usage:      derive
Public Key Object; EC  EC_POINT 256 bits
  EC_POINT:   0441xxxxxx
  EC_PARAMS:  0608xxxxxx
  label:      KEY MAN pubkey
  ID:         03
  Usage:      none
. . . . .
```

Now if we could fix that "Usage: none" for `EC KEY MAN pubkey`...?

Oh, and who/how/when would fix this `SC_ASN1_BIT_STRING_NI` problem in the master?

**Comment by mouse07410** [comment]
Confirm fixed by #1032 . Thanks!
